tumblr/kanvas-ios
# Adds cache to store expensive frame data

This adds a cache which `EditorViewController` can use to store expensive values about an image. For now, that consists of the image's dominant color data used by `EditorViewController.addCarouselDefaultColors()`. This dramatically speeds up switching between images. Eventually, I think we could investigate speeding up the dominant color process, but we don't need to redo this when switching between images.

## Testing

* Enable the `Multiple Exports`, `Camera Metal` and `Metal Filters` flags.
* Add multiple images
* Switch between images
* Notice the speed increase

## Profiling

### Before
<img width="704" alt="Screen Shot 2021-02-12 at 10 25 54 AM" src="https://user-images.githubusercontent.com/3250/107801072-d9788c00-6d1c-11eb-9e15-4aece182db87.png">

A single frame switch takes > 1s:

<img width="787" alt="Screen Shot 2021-02-12 at 10 26 36 AM" src="https://user-images.githubusercontent.com/3250/107801151-f614c400-6d1c-11eb-98af-48668d91178d.png">

### After

Much shorter durations seen in the graph:

<img width="605" alt="Screen Shot 2021-02-12 at 10 24 25 AM" src="https://user-images.githubusercontent.com/3250/107801089-e09f9a00-6d1c-11eb-8791-8e6688e4cbf6.png">

A single frame switch takes ~ 100ms:

<img width="718" alt="Screen Shot 2021-02-12 at 10 25 17 AM" src="https://user-images.githubusercontent.com/3250/107801131-f01ee300-6d1c-11eb-84bc-e677833d6338.png">

### Time Profiler Traces

[BeforeCache.trace.zip](https://github.com/tumblr/kanvas-ios/files/5973421/BeforeCache.trace.zip)
[AfterCache.trace.zip](https://github.com/tumblr/kanvas-ios/files/5973423/AfterCache.trace.zip)
